roxie/udplib/udptrr.cpp

```diff
@@ -469,81 +448,123 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+                    bool anyCanSend = false;
+                    for (UdpSenderEntry *finger = pendingRequests; finger != nullptr; finger = finger->nextSender)
+                    {
+                        if (pendingPermits.length()>=udpMaxPendingPermits)
+                            break;
+                        if (!slots) // || slots<minSlotsPerSender)
+                        {
+                            timeout = 1;   // Slots should free up very soon!
+                            break;
+                        }
+                        // If requester would not be able to send me any (because of the ones in flight) then wait
+                        // Note that the sender may still decide NOT to send any because all it has are still believed to be in-flight and not yet timed out
+                        // In this case I end up with a constant stream of "can I send? Yes? Shan't" until resendTimeout has expired.
+                        // Should be unusual though as (a) implies no more data to send EXCEPT the missing packet (b) resolves once timeout is expired, which should be fairly short (c) implies packet went missing which should be fairly rare
+
+                        // There may be a problem with this logic. Sender won't request permission unless some had timed out. So how is the deadlock ever broken? By canSendAny accounting for timeout...
+                        // Also really I'm able to send as soon as the oldest "in flight" packet arrives, but sender waits until it has timed out before asking if it can send
+
+                        if (finger->canSendAny())
+                        {
+                            unsigned requestSlots = slots;
+                            if (requestSlots>maxSlotsPerSender)
+                                requestSlots = maxSlotsPerSender;
+                            okToSend(finger, requestSlots);
+                            slots -= requestSlots;
+                            if (timeout > udpRequestToSendAckTimeout)
+                                timeout = udpRequestToSendAckTimeout;
+                            anyCanSend = true;
+                        }
+                        else
+                        {
+                            if (udpTraceFlow)
+                            {
+                                StringBuffer s;
+                                DBGLOG("Sender %s can't be given permission to send yet as resend buffer full", finger->dest.getIpText(s).str());
+                            }
+                        }
+                    }
+                    if (pendingRequests && !pendingPermits && !anyCanSend)
+                    {
+                        if (udpTraceFlow)
+                        {
+                            StringBuffer s;
+                            DBGLOG("All senders blocked by resend buffers");
+                        }
+                        timeout = 1; // Hopefully one of the senders should unblock soon
```

Comment:
I'm not sure if this should be if all or any are blocked.  Makes no difference when n=1 though.

Reply:
Discussed